modules/dnn/src/op_inf_engine.cpp

```diff
@@ -828,9 +843,100 @@ void InfEngineBackendNet::addBlobs(const std::vector<Ptr<BackendWrapper> >& ptrs
     }
 }
 
-void InfEngineBackendNet::forward()
+void InfEngineBackendNet::InfEngineReqWrapper::makePromises(const std::vector<Ptr<BackendWrapper> >& outsWrappers)
 {
-    infRequest.Infer();
+    auto outs = infEngineWrappers(outsWrappers);
+    outProms.clear();
+    outProms.resize(outs.size());
+    outsNames.resize(outs.size());
+    for (int i = 0; i < outs.size(); ++i)
+    {
+        outs[i]->promPtr = &outProms[i];
+        outsNames[i] = outs[i]->dataPtr->name;
+    }
+}
+
+void InfEngineBackendNet::forward(const std::vector<Ptr<BackendWrapper> >& outBlobsWrappers,
+                                  bool isAsync)
+{
+    // Look for finished requests.
+    Ptr<InfEngineReqWrapper> reqWrapper;
+    for (auto& wrapper : infRequests)
+    {
+        if (wrapper->isReady)
+        {
+            reqWrapper = wrapper;
+            break;
+        }
+    }
+    if (reqWrapper.empty())
+    {
+        reqWrapper = Ptr<InfEngineReqWrapper>(new InfEngineReqWrapper());
+        reqWrapper->req = netExec.CreateInferRequest();
+        infRequests.push_back(reqWrapper);
+
+        InferenceEngine::BlobMap inpBlobs, outBlobs;
+        for (const auto& it : cnn.getInputsInfo())
+        {
+            const std::string& name = it.first;
+            auto blobIt = allBlobs.find(name);
+            CV_Assert(blobIt != allBlobs.end());
+            inpBlobs[name] = isAsync ? cloneBlob(blobIt->second) : blobIt->second;
+        }
+        for (const auto& it : cnn.getOutputsInfo())
+        {
+            const std::string& name = it.first;
+            auto blobIt = allBlobs.find(name);
+            CV_Assert(blobIt != allBlobs.end());
+            outBlobs[name] = isAsync ? cloneBlob(blobIt->second) : blobIt->second;
+        }
+        reqWrapper->req.SetInput(inpBlobs);
+        reqWrapper->req.SetOutput(outBlobs);
+    }
+    if (isAsync)
+    {
+        // Copy actual data to infer request's input blobs.
+        for (const auto& it : cnn.getInputsInfo())
+        {
+            const std::string& name = it.first;
+            auto blobIt = allBlobs.find(name);
+            Mat srcMat = infEngineBlobToMat(blobIt->second);
+            Mat dstMat = infEngineBlobToMat(reqWrapper->req.GetBlob(name));
+            srcMat.copyTo(dstMat);
+        }
+
+        // Set promises to output blobs wrappers.
+        reqWrapper->makePromises(outBlobsWrappers);
+
+        InferenceEngine::IInferRequest::Ptr infRequestPtr = reqWrapper->req;
+        infRequestPtr->SetUserData(reqWrapper.get(), 0);
+
+        infRequestPtr->SetCompletionCallback({
+            [](InferenceEngine::IInferRequest::Ptr request, InferenceEngine::StatusCode status)
+            {
+                CV_Assert(status == InferenceEngine::StatusCode::OK);
```

Comment:
Probably exceptions are not expected to be escaped from this callback.

```
try
{
    ... pre-checks ...

    for (... outputs ...)
    {
        try {
            wrapper->outProms[i].set_value(m.clone());
        }
        catch(const std::future_error&)
        {
            CV_LOG_ERROR(NULL, "DNN: Exception occured during async inference: promize's std::future_error");
            // probably user-side logic error, ignore?
        }
        catch(...)
        {
            try {
                wrapper->outProms[i]..set_exception(e);
            } catch(...) {  // set_exception() may throw too (see https://en.cppreference.com/w/cpp/thread/promise/set_exception)
                CV_LOG_ERROR(NULL, "DNN: Exception occured during async inference exception propagation");
                // how to recover?
            }
        }
    }
}
catch (...)
{
    std::exception_ptr e = std::current_exception();
    for (... not updated promises ...)
    {
        try {
            wrapper->outProms[i]..set_exception(e);
        } catch(...) {  // set_exception() may throw too (see https://en.cppreference.com/w/cpp/thread/promise/set_exception)
            CV_LOG_ERROR(NULL, "DNN: Exception occured during async inference exception propagation");
            // how to recover?
        }
    }
}
// finally in all cases, we are done here
wrapper->isReady = true;
```

Reply:
Removed it, thanks!